File: Data Structures & Algorithms/target-sum/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number}
     */
    findTargetSumWays(nums, target) {

        const dp = new Map();

        function rec(i,currSum){
            const memoKey = [i,currSum].toString()

            if(dp.has(memoKey))
                return dp.get(memoKey)

            if(i == nums.length)
                return currSum == target ? 1 : 0

            dp[memoKey] =  (
                rec(i+1,currSum + nums[i]) + 
                rec(i+1,currSum - nums[i])
            )

            return dp[memoKey]
        }

        return rec(0,0)
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        s = sum(nums)

        if s%2 == 1:
            return False

        dp = {}

        def rec(i=0,reqSum=(s//2)):
            if (i,reqSum) in dp:
                return dp[(i,reqSum)]

            if reqSum == 0:
                return True
            if i == len(nums):
                return False

            dp[(i,reqSum)]= rec(i+1,reqSum - nums[i]) or \
            rec(i+1,reqSum)

            return dp[(i,reqSum)]


        return rec()
